# Assignment 2
- Donwload the credit dataset 'Credit.csv' from  https://github.com/vahidpartovinia/ycbs255/
- Create the attribute matrix  'X' as a subset of the data that includes only the qualitative attributes 'Income', 'Limit', 'Rating', 'Cards', 'Age', and 'Education'
- Create a discrete response variable, say 'y' by transforming 'Balance' to a binary output, which equals 1 if 'Balance > 1500' and equals 0 otherwise.
- Fit 1- logistic regression, 2- linear discriminant, and 3- quadratic discriminant on the binary Balance as the output variable. Use 'Income', 'Limit', 'Rating', 'Cards', 'Age', and 'Education' as the input variables. 
- Find the probability of (Balance > 1500), for x1 and x2 below, using these three methods. Compare these probabilities and comment.


x1= 'Income' = 63, 'Limit' = 8100, 'Rating' = 600, 'Cards' = 4, 'Age' = 30, 'Education' =14 

x2= 'Income' = 186, 'Limit' = 13414, 'Rating' = 950, 'Cards' = 2, 'Age' = 41, 'Education' =13 

# Submission note
Please fill this jupyter notebook. Extract the pdf file as follows. On  Jupyter manue go to File/Print Preview, then on Browser menu go to File/Print. 

## Only PDF  Submissions will be graded


In [1]:
import pandas as pd 
path='data/'
filename=path+'Credit.csv'
credit=pd.read_csv(filename)
X=credit[['Income','Limit','Rating','Cards','Age','Education']]

In [2]:
X.head()

Income  Limit  Rating  Cards  Age  Education
0   14.891   3606     283      2   34         11
1  106.025   6645     483      3   82         15
2  104.593   7075     514      4   71         11
3  148.924   9504     681      3   36         11
4   55.882   4897     357      2   68         16

In [3]:
import numpy as np
y=np.zeros((400))
y[credit['Balance']>1500]=1

In [4]:
from sklearn.linear_model import LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
from sklearn.metrics import confusion_matrix

In [5]:
lr=LogisticRegression()
lr.fit(X,y)
y_pred = lr.predict(X)
print(confusion_matrix(y, y_pred))

[[390   1]
 [  6   3]]


Comment:Ture negatives 390, false negatives 1, false postives 6, ture positives 3, Logistic Regression has a good performace for dominanat class. The accuracy: 393/400, the recall rate:3/9. This model doesn't satisfy if the inferior class(Balance_1500) is significant. As Losistic Regression models the log-odds rate of data alpha with a linear model(w0+w1*x1+..+Wm*Xm), alpha=w0+w1*x1+..+Wm*Xm, it pay less attention to the inferior class.

In [6]:
lda=LinearDiscriminantAnalysis()
lda.fit(X,y)
y_pred = lda.predict(X)
print(confusion_matrix(y, y_pred))

[[383   8]
 [  4   5]]


Comment: LDA has better performace in inferior class than logistic regression,with recall rate:5/9.  LDA also has a good performance in dominant class(balance<=1500), the total accuracy is 388/400. It seems that LDA pay more atten to inferior class than Logistic Regression. This is possible as LDA is a generative model which estimates p(x|y) for each class(two classes here), then caculates p(y|x) by Bayes rules. Thus LDA is possible to reach a better performance in inferior class than Logistic Regression.

In [7]:
qda=QuadraticDiscriminantAnalysis()
qda.fit(X,y)
y_pred = qda.predict(X)
confusion_matrix(y, y_pred)

array([[390,   1],
       [  2,   7]], dtype=int64)

Comment: QDA has excellent performace for both superior and inferior classes.The total accuracy is 397/400, and the recall rate is 7/9. 
QDA is also a generative model, but without the strong assumption that two classes share the same covariance maxtrix as LDA does. As a result,
QDA has more parameters to estimate and neeeds more computational resources, but it is more flexible with quadratic decision boundary, thus is able to abtain a beeter performance than LDA.

In [8]:
x1=np.array([63,8100,600,4,30,14]).reshape(1,6)
x2=np.array([186,13414,950,2,41,13]).reshape(1,6)

In [9]:
print(lr.predict_proba(x1))
print(lr.predict_proba(x2))
print(lr.predict(x1))
print(lr.predict(x2))

[[ 0.93227393  0.06772607]]
[[ 0.1729211  0.8270789]]
[ 0.]
[ 1.]


The probability of Balance>1500 for x1 and x2 are 0.06772607 and 0.8270789,respectively. And the balace is subject to 0 for x1, and 1 for x2.  

In [10]:
print(lda.predict_proba(x1)[0][1])
print(lda.predict_proba(x2)[0][1])

0.058554280954
0.992788010925


In [11]:
print(qda.predict_proba(x1)[0][1])
print(qda.predict_proba(x2)[0][1])

1.24419206611e-09
0.999216942248


As shown above, the results of classfication for x1 and x2 are identical in three models. x1 belongs to class 0, and x2 is  class 1.But the probabilty of Balance>1500 are different. As analysed before, QDA has highest accuracy in prediction for the calss 1(balance>1500), followed by LDA,then LR. 
Here,the probability of Balance>1500 for x2 is also QDA>LDA>LR>0.5. QDA also has a larger probability of Balance<=1500 for x1. 

In [25]:
credit['Balance_b'] =y

In [26]:
credit.head()

Unnamed: 0   Income  Limit  Rating  Cards  Age  Education  Gender Student  \
0           1   14.891   3606     283      2   34         11    Male      No   
1           2  106.025   6645     483      3   82         15  Female     Yes   
2           3  104.593   7075     514      4   71         11    Male      No   
3           4  148.924   9504     681      3   36         11  Female      No   
4           5   55.882   4897     357      2   68         16    Male      No   

  Married  Ethnicity  Balance  Balance_b  
0     Yes  Caucasian      333        0.0  
1     Yes      Asian      903        0.0  
2      No      Asian      580        0.0  
3      No      Asian      964        0.0  
4     Yes  Caucasian      331        0.0

In [28]:
import statsmodels.formula.api as smf
lr2=smf.Logit.from_formula(formula='Balance_b~Income+Limit+Rating+Cards+Age+Education', data=credit).fit()
lr2.summary2()

Optimization terminated successfully.
         Current function value: 0.040712
         Iterations 11


<class 'statsmodels.iolib.summary2.Summary'>
"""
                        Results: Logit
==============================================================
Model:              Logit            No. Iterations:   11.0000
Dependent Variable: Balance_b        Pseudo R-squared: 0.622  
Date:               2018-04-17 14:17 AIC:              46.5694
No. Observations:   400              BIC:              74.5096
Df Model:           6                Log-Likelihood:   -16.285
Df Residuals:       393              LL-Null:          -43.046
Converged:          1.0000           Scale:            1.0000 
--------------------------------------------------------------
              Coef.   Std.Err.    z    P>|z|   [0.025   0.975]
--------------------------------------------------------------
Intercept    -16.8495   6.0477 -2.7861 0.0053 -28.7028 -4.9962
Income        -0.0840   0.0411 -2.0431 0.0410  -0.1646 -0.0034
Limit          0.0019   0.0036  0.5204 0.6028  -0.0051  0.0089
Rating         0.0125   0.0526  0.2384 0.8116  -0.0906  0.1156
Cards          0.1227   0.4161  0.2948 0.7682  -0.6929  0.9382
Age           -0.0204   0.0313 -0.6536 0.5133  -0.0817  0.0408
Education     -0.0747   0.1420 -0.5263 0.5987  -0.3531  0.2036
==============================================================

"""

Fit logistic regression with statsmodels, the result shows that income is the msot significant variable(|Z|>2,P<0.05) for the category of Balance when setting 'Balance>1500' as the boundary of two classes .